src/libcore/alloc/mod.rs

```diff
@@ -0,0 +1,332 @@
+    //                          allocator" and link it at "valid location".
+    ///
+    /// If the allocation _does_ move, it's the responsibility of the allocator
+    /// to also move the data from the previous location to the new location.
+    Unspecified,
+    /// The address of the new memory must not change.
+    ///
+    /// If the allocation would have to be moved to a new location to fit, the
+    /// reallocation request will fail.
+    InPlace,
+}
+use ReallocPlacement::*;
+
+/// An implementation of `AllocRef` can allocate, grow, shrink, and deallocate arbitrary blocks of
+/// data described via [`Layout`][].
+///
+/// `AllocRef` is designed to be implemented on ZSTs, references, or smart pointers because having
+/// an allocator like `MyAlloc([u8; N])` cannot be moved, without updating the pointers to the
+/// allocated memory.
+///
+/// Unlike [`GlobalAlloc`][], zero-sized allocations are allowed in `AllocRef`. If an underlying
+/// allocator does not support this (like jemalloc) or return a null pointer (such as
+/// `libc::malloc`), this case must be caught. [`Layout::dangling()`][] then can be used to create
+/// an aligned `NonNull<u8>`.
+///
+///
+/// ### Currently allocated memory
+///
+/// Some of the methods require that a memory block be *currently allocated* via an allocator. This
+/// means that:
+///
+/// * the starting address for that memory block was previously returned by [`alloc`], [`grow`], or
+///   [`shrink`], and
+///
+/// * the memory block has not been subsequently deallocated, where blocks are either deallocated
+///   directly by being passed to [`dealloc`] or were changed by being passed to [`grow`] or
+///   [`shrink`] that returns `Ok`. If `grow` or `shrink` have returned `Err`, the passed pointer
+///   remains valid.
+///
+/// [`alloc`]: AllocRef::alloc
+/// [`grow`]: AllocRef::grow
+/// [`shrink`]: AllocRef::shrink
+/// [`dealloc`]: AllocRef::dealloc
+///
+/// ### Memory fitting
+///
+/// Some of the methods require that a layout *fit* a memory block. What it means for a layout to
+/// "fit" a memory block means (or equivalently, for a memory block to "fit" a layout) is that the
+/// following conditions must hold:
+///
+/// * The block's starting address must be aligned to [`layout.align()`], and
+///
+/// * The block's size must fall in the range `use_min ..= use_max`, where:
+///    - `use_min` is [`layout.size()`], and
+///    - `use_max` is the allocation size that was returned.
+///
+/// [`layout.align()`]: Layout::align
+/// [`layout.size()`]: Layout::size
+///
+/// ### Notes
+///
+///  * the size of the layout most recently used to allocate the block is guaranteed to be in the
+///    range `use_min ..= use_max`,
+///
+///  * if a layout `k` fits a memory block (denoted by `ptr`) currently allocated via an allocator
+///    `a`, then it is legal to use that layout to deallocate it, i.e.,
+///    [`a.dealloc(ptr, k);`][`dealloc`], and
```

Comment:
I don't know if I get your point here. How the alignment of an allocation can differ? The *memory fitting* section says that "the block's starting address must be aligned to `layout.align()`" and `alloc` must ensure, that the returned block is aligned to `layout.align()`.

Reply:
The returned pointer could be aligned even more than requested. For a concrete example, a bump pointer allocator might grab its block of memory directly from `mmap`, getting a page-aligned block (say to 4096). Then, upon calling
```rust
let ptr = allocator.alloc(Layout::from_size_align(2, 2)).unwrap().0;
```
the allocator will return a pointer to the first two bytes of that block. This pointer is aligned to the alignment I requested (2), but it is also aligned much higher, at least up to 4096, since it is the same pointer that was directly returned by `mmap` as the beginning of the block. The layout `Layout::from_size_align(2,4096)` also fits that memory block because the actual pointer to the block happens to have its low 12 bits cleared. (Even more commonly, `malloc` is required to have a minimum allocation on all of its allocations of usually 8 or 16.) According to the documentation here, it is then safe for me to call
```rust
allocator.dealloc(ptr, Layout::from_size_align(2,4096));
```
yet the documentation on `dealloc` itself disagrees.

Reply:
Ah sure, I'll adjust the documentation of `dealloc`, that `layout.align()` has to be the same as used in `alloc`. Otherwise you wouldn't need a `Layout` in `dealloc` but only a `size`.

Reply:
Wait, actually that's how it is written down.
Surely the returned pointer can be overaligned, but you don't have to care. The supposed way to do is the way how it is mentioned in `dealloc`.

Reply:
I changed the memory fitting secion to

```rust
/// * The block must be allocated with the same alignment as [`layout.align()`], and
```

Reply:
Great, that sounds good to me.

(It might be possible to loosen the requirements later, but that might introduce extra and unnecessary complexity to allocator implementations, doesn't seem to have a compelling use case, and, if desired, probably warrants more thorough discussion by the WG.)